tests/tests/test_filetransfer.py

```diff
@@ -0,0 +1,190 @@
+        "https://%s/api/management/v1/deviceconnect"
+        % get_container_manager().get_mender_gateway()
+    )
+    upload_url = "%s/devices/%s/upload" % (deviceconnect_url, devid)
+    return requests.put(upload_url, verify=False, headers=authtoken, files=files)
+
+
+def md5sum(fname):
+    hash_md5 = hashlib.md5()
+    with open(fname, "rb") as f:
+        for chunk in iter(lambda: f.read(4096), b""):
+            hash_md5.update(chunk)
+    return hash_md5.hexdigest()
+
+
+@pytest.mark.usefixtures("standard_setup_one_client")
+class TestFileTransfer(MenderTesting):
+    """Tests the file transfer functionality"""
+
+    def test_filetransfer(self, standard_setup_one_client):
+        """Tests the file transfer features
+        """
+        # accept the device
+        devauth.accept_devices(1)
+
+        # list of devices
+        devices = list(
+            set([device["id"] for device in devauth.get_devices_status("accepted")])
+        )
+        assert 1 == len(devices)
+
+        # wait for the device to connect via websocket
+        auth = authentication.Authentication()
+        wait_for_connect(auth, devices[0])
+
+        # device ID and auth token
+        devid = devices[0]
+        authtoken = auth.get_auth_token()
+
+        # download a file and check its content
+        path = "/etc/mender/mender.conf"
+        r = download_file(path, devid, authtoken)
+        assert r.status_code == 200, r.json()
+        assert "ServerURL" in str(r.content)
+        assert (
+            r.headers.get("Content-Disposition") == 'attachment; filename="mender.conf"'
+        )
+        assert r.headers.get("Content-Type") == "application/octet-stream"
+        assert r.headers.get("X-Men-File-Gid") == "0"
+        assert r.headers.get("X-Men-File-Uid") == "0"
+        assert r.headers.get("X-Men-File-Mode") == "600"
+        assert r.headers.get("X-Men-File-Path") == "/etc/mender/mender.conf"
+        assert r.headers.get("X-Men-File-Size") != ""
+
+        # wrong request, path is relative
+        path = "relative/path"
+        r = download_file(path, devid, authtoken)
+        assert r.status_code == 400, r.json()
+        assert r.json().get("error") == "bad request: path: must be absolute."
+
+        # wrong request, no such file or directory
+        path = "/does/not/exist"
+        r = download_file(path, devid, authtoken)
+        assert r.status_code == 400, r.json()
+        assert "/does/not/exist: no such file or directory" in r.json().get("error")
+
+        try:
+            # create a 40MB random file
+            f = NamedTemporaryFile(delete=False)
+            for i in range(40 * 1024):
+                f.write(os.urandom(1024))
+            f.close()
+
+            # upload the file
+            r = upload_file(
+                "/tmp/random.bin",
+                open(f.name, "rb"),
+                devid,
+                authtoken,
+                mode="600",
+                uid="0",
+                gid="0",
```

Comment:
Suggestion: Use some random mode/uid/gid values here, which will test that both the upload and the download of these work.